src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,5 +353,52 @@ private BoundExpression RewriteStringConcatInExpressionLambda(CSharpSyntaxNode s
 
             return new BoundBinaryOperator(syntax, operatorKind, loweredLeft, loweredRight, default(ConstantValue), method, default(LookupResultKind), type);
         }
+
+        /// <summary>
+        /// Checks whether the expression represents a boxing conversion of a value type.
+        /// If it does, it returns a synthesized ToString call for that value type instance.
+        /// Otherwise, it returns the original expression.
+        /// </summary>
+        private BoundExpression ConvertExpressionToStringIfValueType(CSharpSyntaxNode syntax, BoundExpression expr)
+        {
+            // The lowered expressions coming in will be boxing conversions if the input is a value type.
+            // When passed to String.Concat, the resulting object will have ToString called on it.
+            // We can optimize away the boxing allocation by just calling ToString on the value type directly.
```

Comment:
It looks like a breaking change and a deviation from C# spec in some cases. If the value type is mutable and its method ToString() is mutating, then currently this mutation is not observable because it's made on a boxed copy. With this change it will directly mutate the value, that can have observable effects. This looks like a safe optimization only for known immutable value types like int, Guid, DateTime etc.